src/test/app/AMM_test.cpp

```diff
@@ -4132,6 +4132,341 @@ struct AMM_test : public jtx::AMMTest
         });
     }
 
+    void
+    testSelection()
+    {
+        testcase("Offer/Strand Selection");
+        using namespace jtx;
+        Account const ed("ed");
+        Account const gw1("gw1");
+        auto const ETH = gw1["ETH"];
+        auto const CAN = gw1["CAN"];
+
+        auto prep = [&](Env& env, auto gwRate, auto gw1Rate) {
+            fund(env, gw, {alice, carol, bob, ed}, XRP(2'000), {USD(2'000)});
+            env.fund(XRP(2'000), gw1);
+            fund(
+                env,
+                gw1,
+                {alice, carol, bob, ed},
+                {ETH(2'000), CAN(2'000)},
+                Fund::IOUOnly);
+            env(rate(gw, gwRate));
+            env(rate(gw1, gw1Rate));
+            env.close();
+        };
+
+        for (auto const& rates :
+             {std::make_pair(1.5, 1.9), std::make_pair(1.9, 1.5)})
+        {
+            // Offer Selection
+
+            // Cross-currency payment: AMM has the same spot price quality
+            // as CLOB's offer and can't generate a better quality offer.
+            // The transfer fee in this case doesn't change the CLOB quality
+            // because trIn is ignored on adjustment and trOut on payment is
+            // also ignored because ownerPaysTransferFee is false in this case.
+            // Run test for 0) offer, 1) AMM, 2) offer and AMM
+            // to verify that the quality is better in the first case,
+            // and CLOB is selected in the second case.
+            {
+                std::array<Quality, 3> q;
+                for (auto i = 0; i < 3; ++i)
+                {
+                    Env env(*this);
+                    prep(env, rates.first, rates.second);
+                    std::shared_ptr<AMM> amm;
```

Comment:
Why are you using `shared_ptr` in these tests? Since `amm` is not used anywhere outside this block, it probably makes more sense to use `std::optional`, and `emplace` the object if you want it. I'm not sure if there's a performance benefit either way - it just seems cleaner to me.